File: project_manager.py
#!/usr/local/bin/python

import calendar, time, pickle, cgi, cgitb, os
cgitb.enable()

compose     = lambda m, n                     : lambda x, m=m, n=n: m(n(x))
link        = lambda href, text, id='', cls='': "<a id='%s' class='%s' href='%s'>%s</a>" % (id, cls, href, text)
wrap        = lambda tag, data, rest=''       : "<%s> %s </%s>" % (tag, data, tag)
label       = lambda name, value=''           : "<label for='%s'>%s</label><input name='%s' value='%s' />" % (name, name, name, value)
submit      = lambda                          : "<input type=submit name=submit value='Submit'>"
hidden      = lambda n, v                     : "<input type=hidden name='%s' value='%s' />" % (n, v)
project     = lambda t                        : {'title': t,'members': [], 'overview': "", 'comments': [], 'tasks': []}
loadproject = lambda p                        : pickle.load(open(p, "rb"))
loaduser    = lambda u                        : pickle.load(open(u, "rb"))
tags        = ('a','b','div','h1','h2','h3','img','p','span')
a, b, div, h1, h2, h3, img, p, span = map(lambda tag: lambda str: wrap(tag, str), tags)

def datepicker(d=1, m=1, y=2007):
	year, month, thedate, x, y, z, a, xx, yy = time.localtime()
	ypart = dselect('year', [x for x in range(2007, 2030)], year)
	mpart = dselect('month',[x for x in range(1, 12)], month)
	dpart = dselect('date', [x for x in range(1, 32)], thedate)
	return "%s %s %s" % (ypart, mpart, dpart)

def dselect(name, lst, default=''):
	select = "<select name='%s'>" % name
	for opt in iter(lst):
		if str(opt) == str(default): selected = "selected='selected'"
		else: selected = ''
		select = select + "<option value='%s' %s > %s</option>" % (opt, selected, opt)
	return (select + "</select>")

def saveproject(p):
	pickle.dump(p, open(p['title'] + ".prj", "wb"))
	return p

def saveuser(u):
    pickle.dump(u, open(u['name'] + ".usr", "wb"))
    return u

def print_header():
	print "Content-Type: text/html\n\n"
	print "<html><head><link type='text/css' rel='stylesheet' href='/project_manager/style.css' />",
	print "<script type='text/javascript' src='/project_manager/core.js' />"
	print "<script type='text/javascript' src='/project_manager/app.js' />"
	print "<script type='text/javascript' src='/project_manager/ajax.js' />"
	print "</head><body>"

def print_footer(): print "</div></body></html>"

def flatten(l):
	if type(l) !=  type([]): return [l]
	if l == []: return l
	return flatten(l[0]) + flatten(l[1:])

def saveform(ob):
	keys = ob.keys()
	post = cgi.FieldStorage()
	for x in iter(post):
		if ob.has_key(x): ob[x] = post[x]
	return ob

def makeform(name, show, hide):
	frm = ""
	for x in show: frm = frm + label(x) + "<br>"
	for x in hide: frm = frm + "<input type='hidden' name='%s' />" % x
	return frm

def show_all_projects():
	projects = os.listdir(".")
	projects = [project(p) for p in projects if os.path.isfile(p) and os.path.splitext(p)[1] in [".prj"]]
	print "<ul id=projectlist><li id=createproject ><a href='?addproject=1'>Create Project</a></li>"
	for p in iter(projects):
		ptitle = os.path.splitext(p['title'])[0]
		print "<li><a href=\"?showproject=%s\">%s</a></li>" % (p['title'], ptitle)
	print "</ul><div id='cnt'>"

def showproject( p , all='normal'):
	print h1(p['title']),"<a href='?showall=%s'>Show All</a> | <a href='?showproject=%s.prj'>Show incomplete</a><a href='?addtask=%s' class='addtask'>Add a task</a>" % (p['title'], p['title'], p['title']), h2("Total hours: %.2f" % (get_project_time(p)))
	print "<div id='members'>"
	if not p.has_key('members'): p['members'] = []
	print h3(" Team members")
	for i in iter(p['members']): print i['name'],", "
	print link('?project=' + p['title'] + '&addmember=1', 'Add a team member')
	print "</div>"
	print " <div id=tasklist>"
	for i in iter(p['tasks']):
		if not i.has_key('status'): i['status'] = 'incomplete'
		if i['status'] == 'completed' and all == 'normal' : continue
		print "<div class=taskline>"
		print " <a class='%s' href='?project=%s&addwork=%s'>Add hours</a>" % (i['status'], p['title'], i['title'])
		print " <a class='%s' href='?project=%s&setstatus=%s'>Complete</a>" % (i['status'], p['title'], i['title'])
		print "<span class='name'>%s</span>" % i['person'],
		print " <span class='task'>%s</span>" % i['title']
		print "</div>"
		print "<div class=timedata>"
		for x in iter(i['time']):
			print "<div>"
			print "<form style='display:inline' action='' method=post >"
			print "Date <span class=date>%s</span> Duration <span class=hours>%.2f</span>" % (x['date'], x['hours'])
			print "<a class=changehours project='%s' task='%s' date='%s' hours='%2.f' href='#' >Change</a>" % (p['title'], i['title'], x['date'], x['hours'])
			print hidden('savechangehours', 1), hidden('project', p['title']), hidden('task', i['title'])
			print "</form>"
			print "</div>"
		print "</div>"
	print "</div>"

def settaskstatus(p, t):
	task = gettask(p, t)
	task['status'] = 'completed'
	ts = [x for x in p['tasks'] if x['title'] != t]
	ts.append(task)
	p['tasks'] = ts
	return saveproject(p)

def gettask(p, t):
	return [x for x in p['tasks'] if x['title'] == t][0]

def savemember(project, person):
	if not project.has_key('members'): project['members'] = []
	names = [x['name'] for x in project['members']]
	r = False
	if len(names) > 0: r = reduce(lambda x, y: (x == person) or y, names)
	if r != person: project['members'].append({'name':person})
	return saveproject(project)

def savetask(project, title, person, status='incomplete' ):
	project['tasks'].append( {'person':person, 'title':title, 'status':status, 'time':[]} )
	return saveproject(project)

def savework(project, title, year, month, day, hours):
	date = "%s-%s-%s" % (day, month, year)
	task  = gettask(project, title)
	hours = float(hours)
	tasks = [x for x in project['tasks'] if x['title'] != title]
	samedatelist = [x for x in task['time'] if x['date'] == date]
	samedate = ''
	if len(samedatelist): samedate = samedatelist[0]
	notsamedate = [x for x in task['time'] if x['date'] != date]

	if samedate:
		samedate['hours'] = float(samedate['hours']) + hours
		notsamedate.append(samedate)
		task['time'] = notsamedate
	else:
		task['time'].append( {'date':"%s-%s-%s" % (year, month, day), 'hours':hours} )

	tasks.append( task )
	project['tasks'] = tasks
	return saveproject(project)

def get_project_time(project):
	times = map(lambda x: x['time'], project['tasks'])
	total = 0
	if len(times) > 0:
		hourlist = map(lambda x: x['hours'], flatten(times))
		if len(hourlist) > 0: total = reduce(lambda x, y: x + y, hourlist)
	return total

def showtasks(p):
	for t in p['tasks']:
		print t['title'], "(", t['person'], ")"

def addmember(p):
	print "<p>Project: <b>%s</b></p>" % p
	print "<form name=addmember method=post action=''>"
	print "<fieldset><legend>Add a new member to the %s project</legend>" % p
	print label('person'), submit(), hidden('project', p), hidden('savemember','1')
	print "</form>"

def addtask(p):
	proj = loadproject(p + ".prj")
	people = [m['name'] for m in proj['members']]
	print "<p>Project: <b>%s</b></p>" % p
	print "<fieldset><legend>Add task to %s p</legend>" % p
	print "<form name=addtask method=post action=''>"
	print label('task'), dselect('person', people), submit(), hidden('project', p), hidden('savetask','1')
	print "</fieldset></form>"

def addwork(p, t, change=0):
	print "<p>Project: <b>%s</b></p>" % p
	print "<form name=add_work method=post action=''>"
	print "<fieldset><legend>Task: %s</legend>" % t
	print datepicker(), label('hours'), submit(), hidden('project', p), hidden('task', t), hidden('savework','1')
	print "</form>"

def changehours(p, t, h):
	print "<p>Project: <b>%s</b></p>" % p
	print "<form name=add_work method=post action=''>"
	print "<fieldset><legend>Task: %s</legend>" % t
	print datepicker(), label('hours'), submit(), hidden('project', p), hidden('task', t), hidden('savework','1')
	print "</form>"

def addproject():
	print " <form name=addproject method=post action='' />"
	print label('name'),"<br>"
	print "<label for=overview>About</label><textarea name='overview' class=overview></textarea><br>"
	print submit(), hidden('saveproject','1')
	print "</form>"

def check_post():
	post = cgi.FieldStorage()
	if post.has_key('showproject') : showproject(loadproject(post['showproject'].value))
	if post.has_key('showall')     : showproject(loadproject(post['showall'].value + ".prj"),'showall')
	if post.has_key('addproject')  : p = addproject()
	if post.has_key('saveproject') : showproject(saveproject(project(post['name'].value)))
	if post.has_key('addmember')   : addmember(post['project'].value)
	if post.has_key('addwork')     : addwork(post['project'].value, post['addwork'].value)
	if post.has_key('addtask')     : addtask(post['addtask'].value)

	if post.has_key('savechangehours'):
		p                                      = post['project'].value
		p                                      = loadproject(p + '.prj')
		t                                      = post['task'].value
		d                                      = post['date'].value
		h                                      = float(post['hours'].value)
		task                                   = gettask(p, t)
		times                                  = task['time']
		leavethem                              = [x for x in times if x['date'] != d]
		leavethem.append({'date':d,'hours':h})
		task['time']                           = leavethem
		leavethem                              = [x for x in p['tasks'] if x['title'] != t]
		leavethem.append(task)
		p['tasks'] = leavethem
		showproject(saveproject(p))

	if post.has_key('savemember'):
		p, m = [post[x].value for x in ['project','person']]
		return showproject(savemember(loadproject(p + ".prj"), m))

	if post.has_key('savetask'):
		p, t, person = [post[x].value for x in ['project','task','person']]
		return showproject(savetask(loadproject(p + ".prj"), t, person))

	if post.has_key('savework'):
		p, t, y, m, d, hrs = [post[x].value for x in ['project','task','year','month','date','hours']]
		return showproject(savework(loadproject(p + ".prj"), t, y, m, d, hrs))

	if post.has_key('setstatus'):
		p, t = [post[x].value for x in ['project','setstatus']]
		return showproject(settaskstatus(loadproject(p + ".prj"), t))

def main():
	print_header()
	show_all_projects()
	check_post()
	print_footer()

main()

File: projectmanager.rb
#!/usr/bin/ruby

require 'cgi'
require 'cgi/session'

cgi = CGI.new
print cgi.header("type" => "text/html")

def compose(f,g)
    lambda{ |x| f(g(x)) }
end

def xtag (kind,title)
	"<#{kind}>text</#{kind}>"
end

def htmltag
	tags = ['div','p','h1','h2','h3','script']
	tags.each do |t| 
		#{t = lambda { | atts | "<#{t} #{atts.each {|k,a| print k; print a}} > </{t}>" }} 
	end	
end
htmltags()

def br
	print "<br />"
end

def link(href,text,id='',cls='')
    print "<a id='#{id}' class='#{cls}' href='#{href}'>#{text}</a>"
end

def wrap(tag,data)
    print "<#{tag}> #{data} </#{tag}>"
end

def label(n,v='')
    print "<label for='#{n}'>#{n}</label><input name='#{n}' value='#{v}' />"
end

def submit
    print "<input type=submit name=submit value='Submit'>"
end

def hidden(n,v='')
    "<input type=hidden name='#{n}' value='#{v}' />"
end

def get_project_time(project)
	times = project['tasks'].map {|x| x['time']}
	total = 0
	if len(times) > 0
		hourlist = times.flatten!.map {|x| x['hours']}
        total = hourlist.inject {|x,y| x + y} if len(hourlist) > 0
	end
	total
end

def datepicker(d=0,m=0,y=0)
	t = Time.new
	ypart = dselect('year',(2007 .. 2030).to_a, t.year)
	mpart = dselect('month',(1 .. 12).to_a,t.month)
	dpart = dselect('date',(1 .. 31),t.day)
	"#{ypart} #{mpart} #{dpart}"
end

def dselect(name, lst, default='')
	select = "<select name='#{name}'>"
	lst.each {|opt|
		selected = (opt==default) ? "selected='selected'" : selected = ''
		select = select + "<option value='#{opt}' #{selected} >#{opt}</option>"
    }
	select + "</select>"
end

def gettask(project_title, t)
	 p = loadproject(project_title)
	 task = p['tasks'].find {|x| x['title'] = t }
	 print task['title']
	 task
end

def newproject(t)
    p = {'title' => t,'members' => [],'overview' => "",'comments' => [],'tasks' => [] }
	saveproject p
end
def loadproject(p)
    File.open("#{p}.prj") {|f| Marshal.load(f) }
end
def addproject()
	print " <form name=addproject method=post action='' />"
	label("name");br();submit(); hidden("saveproject",1)
	print "</form>"
end
def saveproject(p)
	File.open("#{p['title']}.prj",'w'){|f| Marshal.dump(p,f)}
	loadproject(p)
end
def showproject( p , all='normal')
	print "all is #{all}..."
	un = p['tasks'].find {|x| x['status']=='completed'}
	print "<h1>#{p['title']}</h1>"
	print "<p>#{p['overview']}</p>"
	if p['members'] != []
    	if un
            link("?showall=#{p['title']}",     "Show All",        "showall")
            link("?showproject=#{p['title']}", "Show incomplete", "noshowall")
        end
        link("?addtask=#{p['title']}","Add a task","","addtask")
	end
	print "<div id='members'><h3>Team members</h3>"
	p['members'].each {|i| print "#{i} "}
	link("?project=#{p['title']}&addmember=1", "Add a team member")
	print "</div><div id=tasklist>"
	p['tasks'].each {|i|
		i['status'] = 'incomplete' if not i.has_key?('status')
		continue if (i['status'] == 'completed' and all == "")
		print "<div class=taskline>"
		link("?project=#{p['title']}&addwork=#{i['title']}",   "Add hours", i['status'])
		link("?project=#{p['title']}&setstatus=#{i['title']}", "Complete",  i['status'])
		print "<span class='name'>#{i['person']}</span>"
		print " <span class='task'>#{i['title']}</span></div>"
		print "<div class=timedata>"
		i['time'].each {|x|
			print "<div><form style='displayinline' action='' method=post >"
			print "Date <span class=date>#{x['date']}</span> Duration <span class=hours>#{x['hours']}</span>"
			print "<a class=changehours project='#{p['title']}' task='#{i['title']}' date='#{x['date']}' hours='#{x['hours']}' href='#' >Change</a>"
			hidden('savechangehours',1); hidden('project',p['title']); hidden('task',i['title'])
			print "</form></div>"
        }
		print "</div>"
    }
	print "</div>"
end

def print_header
	print "<html><head>"
	print "<link type='text/css' rel='stylesheet' href='/project_manager/style.css' />"
	['core','app','ajax'].each do |j|
        print "<script type='text/javascript' src='/project_manager/#{j}.js' />"
	end
	print "</head><body>"
end

def print_footer
    print "</div></div></body></html>"
end

def saveform(ob,post)
	keys = ob.keys
	post.each {|x| ob[x] = post[x] if ob.has_key?(x)}
	ob
end

def makeform(name,show,hide)
	frm = ""
	show.each {|x| frm += label(x) + "<br>"}
	hide.each {|x| frm += "<input type='hidden' name='#{x}' />"}
	frm
end


def showtasks(p)
	p['tasks'].each {|t| print t['title'] + "(" + t['person'] + ")"}
end

def show_all_projects()
	projects = Dir.entries('.').find_all {|p| p.split(".")[1] == "prj"}
	print "<ul id=projectlist><li id=createproject ><a href='?addproject=1'>Create Project</a></li>"
	projects.each {|p|
		p = p.split(".")[0]
		print "<li><a href='?showproject=#{p}'>#{p}</a></li>"
    }
	print "</ul><div id='cnt'>"
end

def addmember(p)
	print "<p>Project <b>#{p}</b></p>"
	print "<form name=addmember method=post action=''>"
	print "<fieldset><legend>Add a new member to the #{p} project</legend>"
	print label('person')
	print submit()
	print hidden('project',p)
	print hidden('savemember','1')
	print "</form>"
end
def savemember(project,person)
    project['members'] = [] if not project.has_key?('members')
	names = project['members'] << person
	names.uniq!
    project['members'] = names
	saveproject(project)
end


def addtask(p)
	proj = loadproject(p)
	people = proj['members']
	print "<p>Project <b>#{p}</b></p>"
	print "<fieldset><legend>Add task to the #{p} project</legend>"
	print "<form name=addtask method=post action=''>"
	print "Task <input name='task' />"
	print " New team member <select name='person'>"
	people.each {|x| print "<option value=#{x}>#{x}</option>" }
	print "</select> "
	print "<input type=submit name=submit value=submit />"
	print hidden('project',p)
	print hidden('savetask','1')
	print "</fieldset></form>"
end

def savetask(project, title, person, status='incomplete' )
	project = loadproject(project)
	project['tasks'] << {'person' => person, 'title' => title, 'status' => status, 'time' => []}
	print "Project is #{project}<br>"
	project = saveproject(project)
	print "and now Project is #{project}<br>"
end

def settaskstatus(p,t)
	task = gettask(p,t)
	p = loadproject(p)
	task['status'] = 'completed'
	tasks = p['tasks'].find_all {|x| x['title'] != t }
	p['tasks'] = tasks << task
	saveproject(p)
end


def addwork(p,t)
	print "<p>Project: <b>#{p}</b></p>"
	print "<form name=add_work method=post action=''>"
	print "<fieldset><legend>Task: <b>#{t}</b></legend>"
	print datepicker
	print label('hours')
	print submit
	print hidden('project',p)
	print hidden('task',t)
	print hidden('savework','1')
	print "</form>"
end
def savework(project,title,y,m,d,hours)
	date = "#{d}-#{m}-#{y}"
	task  = project['tasks'].find {|x| x['title'] == title }
	tasks = project['tasks'].find_all {|x| x['title'] != title}
	samedate = (task) ? task['time'].find {|x| x['date'] == date} : false
	notsamedate = task['time'].find_all {|x| x['date'] != samedate['date'] }
	if samedate
		samedate['hours'] = samedate['hours'].to_f + hours.to_f
		task['time'] = notsamedate << samedate
	else
		task['time'] << {'date' => date, 'hours' => hours}
    end
	project['tasks'] = tasks << task
	saveproject(project)
end


def check_post(post)
	if post.has_key?('showproject')  
        project = loadproject(post['showproject'])
	end
	if post.has_key?('showall')  
        project = loadproject(post['showall'])  
	end
	if post.has_key?('saveproject')
        project = newproject(post['name'])
	end
	if post.has_key?('addmember')
        project = addmember(post['project'])
	end
	if post.has_key?('addproject')
        project = addproject
    end

	if post.has_key?('addwork')
        project = addwork(post['project'], post['addwork'])
	end
	if post.has_key?('addtask')
        project = addtask(post['addtask'])
    end
	if post.has_key?('savechangehours')
		t,d,h = post['task'],post['date'],post['hours']
		task = gettask(p,t)
		task['time'] = task['time'].find_all {|x| x['date'] != d} << {'date' => d,'hours' => h}
		p['tasks'] = p['tasks'].find_all {|x| x['title'] != t} << task
		project = saveproject(p)
	end

	if post.has_key?('savemember')
		project = savemember(post['project'],post['person'])
	end

	if post.has_key?('savetask')
		project = savetask(post['project'],post['task'],post['person'])
	end

	if post.has_key?('savework')
		p,t,y,m,d,hrs = post['project'],post['task'],post['year'],post['month'],post['date'],post['hours']
		project = savework(p,t,y,m,d,hrs)
	end
	if post.has_key?('setstatus')
		project = settaskstatus(post['project'],post['setstatus'])
	end
	show = post['showall']	
	showproject(project,show)
end

print_header
show_all_projects
check_post(cgi)
print_footer
